### PR TITLE
Favorites and notifications

### PR DESCRIPTION
**Pull request by FacuCristobalGlobant**
Implemented a way to save specific cards as favorites in the Firestore database, and also made it so a notification will show when saving a card.

https://user-images.githubusercontent.com/109153713/192877749-8a9a041a-5650-434b-ba65-6e1e1a5e870f.mp4

Added new tests and adapted the ones already implemented.

Screenshot 2022-09-28 104428

### CODE REVIEW

File: lib/local_notice_service.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  Future<void> showNotification(int id, String title, String body) async {
+    const androidDetail = AndroidNotificationDetails(
+      'HearthstoneApp',
+      'Favorites',
+    );
+    const noticeDetail = NotificationDetails(
+      android: androidDetail,
+    );
+    await _localNotificationsPlugin.show(id, title, body, noticeDetail);
```
Comment (FranciscoGaraventa) [inline lib/local_notice_service.dart]: trailing comma

File: lib/local_notice_service.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    );
+
+    await _localNotificationsPlugin.initialize(initSettings).then((_) {
+      debugPrint('setupPlugin: setup success');
+    }).catchError((Object error) {
+      debugPrint('Error: $error');
+    });
+  }
+
+  Future<void> showNotification(int id, String title, String body) async {
```
Comment (FranciscoGaraventa) [inline lib/local_notice_service.dart]: trailing comma

File: lib/src/data/datasource/local/DAOs/firestore_database.dart [hunk 3]
```diff
@@ -19,6 +19,53 @@ class FirestoreDatabase {
     batch.commit();
   }
 
+  Future<List<Map<String, dynamic>>> getFavorites() async {
+    final List<Map<String, dynamic>> favorites = [];
+    try {
+      final QuerySnapshot snapshot = await _firestore
+          .collection('cards')
+          .where('favorite', isEqualTo: true)
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/local/DAOs/firestore_database.dart]: Trailling comma and add collection to constants

File: lib/src/data/datasource/local/DAOs/firestore_database.dart [hunk 4]
```diff
@@ -19,6 +19,53 @@ class FirestoreDatabase {
...
+    await reference.update({'favorite': true});
+    return true;
+  }
+
+  Future<bool> removeFavorite(String cardId) async {
+    try {
+      await _firestore
+          .collection('cards')
+          .doc(cardId)
+          .update({'favorite': FieldValue.delete()});
```
Comment (FranciscoGaraventa) [inline lib/src/data/datasource/local/DAOs/firestore_database.dart]: add collection to constants and favourite field too

File: lib/src/presentation/bloc/favorites_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      StreamController.broadcast();
+
+  FavoritesBloc({FavoritesRepository? repository}) {
+    favoritesRepository = repository ?? FavoritesRepository();
+  }
+
+  @override
+  Stream<ApiResponse> getCardListStream() {
+    return _favoritesCardListController.stream;
+  }
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/favorites_bloc.dart]: you can replace this with an arrow function

File: lib/src/presentation/bloc/favorites_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  @override
+  Future<void> setFavorite(String cardId) async {
+    favoritesRepository.setFavorite(cardId);
+  }
+
+  @override
+  Future<bool> isFavorite(String cardId) async {
+    return favoritesRepository.isFavorite(cardId);
+  }
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/bloc/favorites_bloc.dart]: you can replace every call with an arrow function

File: lib/src/presentation/widget/card_list_builder.dart [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  final Stream<ApiResponse> stream;
+  final FavoritesBloc favoritesBloc;
+
+  const CardListBuilder(this.stream, this.favoritesBloc);
+
+  @override
+  Widget build(BuildContext context) {
+    return StreamBuilder<ApiResponse>(
+      stream: stream,
+      builder: (context, snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/card_list_builder.dart]: trailling comma and specify context/snapshot types

File: lib/src/presentation/widget/favorite.dart [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+class _FavoriteState extends State<Favorite> {
+  bool _isFavorite = false;
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
+      future: widget.favoritesBloc.isFavorite(widget.card.cardId!),
+      builder: (BuildContext context, AsyncSnapshot<bool> snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/favorite.dart]: trailling comma

File: lib/src/presentation/widget/favorites_list.dart [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  final FavoritesBloc favoritesBloc;
+
+  const FavoritesList(this.favoritesBloc);
+
+  @override
+  Widget build(BuildContext context) {
+    return StreamBuilder<ApiResponse>(
+      key: UniqueKey(),
+      stream: favoritesBloc.getCardListStream(),
+      builder: (context, snapshot) {
```
Comment (FranciscoGaraventa) [inline lib/src/presentation/widget/favorites_list.dart]: Trailling comma and specify context/snapshot types

File: lib/local_notice_service.dart (hunk 1 above)
Comment (FacuCristobalGlobant) [reply lib/local_notice_service.dart]: Added

File: lib/local_notice_service.dart (hunk 2 above)
Comment (FacuCristobalGlobant) [reply lib/local_notice_service.dart]: Added

File: lib/src/data/datasource/local/DAOs/firestore_database.dart (hunk 3 above)
Comment (FacuCristobalGlobant) [reply lib/src/data/datasource/local/DAOs/firestore_database.dart]: Done

File: lib/src/data/datasource/local/DAOs/firestore_database.dart (hunk 4 above)
Comment (FacuCristobalGlobant) [reply lib/src/data/datasource/local/DAOs/firestore_database.dart]: Done

File: lib/src/presentation/bloc/favorites_bloc.dart (hunk 5 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/bloc/favorites_bloc.dart]: Done

File: lib/src/presentation/bloc/favorites_bloc.dart (hunk 6 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/bloc/favorites_bloc.dart]: Done

File: lib/src/presentation/widget/card_list_builder.dart (hunk 7 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/card_list_builder.dart]: Added

File: lib/src/presentation/widget/favorite.dart (hunk 8 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/favorite.dart]: Added

File: lib/src/presentation/widget/favorites_list.dart (hunk 9 above)
Comment (FacuCristobalGlobant) [reply lib/src/presentation/widget/favorites_list.dart]: Done

**Review by FranciscoGaraventa — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Review by Juliana-Dimeglio — APPROVED** [review]

**Pull request merged**